### ISSUE TITLE
Incompatible with hot module reload?

### ISSUE THREAD

**Issue by Danita** [i18next/react-i18next#6]
First of all, thank you for this work, :smile: it really looks a lot easier to use than `react-intl` which I find a little overkill for a simple app.

I have a translated view that goes like this:

```jsx
import React from 'react';
import { translate } from 'react-i18next/lib';

const LoginForm = React.createClass({
	render() {
		const t = this.props.t;
		console.log(t); // <-- we should have the translation fn here
		return (
			<div>
				<h3>{ t('Sign in')}</h3>
				{/*...*/}
			</div>
		);
	}
});

export default translate(['login'])(LoginForm);
```
On first load and manual reload it works as advertised, displaying the translated message. The `console.log(t)` call returns something like `function fixedT(key, options) {...}`, but when I modify something on the component and thus triggering the hot reload module (Webpack), I get `undefined` and an `Uncaught TypeError: t is not a function`.

Is there any way to make react-i18next compatible with HMR?

Thanks!

**Comment by jamuhl** [comment]
will have a look on monday.

**Comment by jamuhl** [comment]
tried to reproduce on our env...but not having an issue with HMR...but might be we got a different configuration plus different versions of webpack.

Do you have a small project i could access to reproduce?

**Comment by Danita** [comment]
Hey Jan, thanks for taking a look at this. I have the latest versions of all dependencies, including Webpack. As of now, I replaced `react-i18next` for `react-intl`, but I think I still have the branch where I tried `react-18next`, so I could upload a repro case.

**Comment by jamuhl** [comment]
never mind...thanks for the reply...closing this for now.

**Comment by sorahn** [comment]
I just ran into this same issue.

I'm using the code as a decorator

```javascript
@translate(['common'])
class Header extends Component {
  static propTypes = {
    t: PropTypes.func
  };

  render () {
    const { t } = this.props

    return <h1>{t('hello_world')}</h1>
  }
}
```

Here's my i18n.js init.

```javascript
import i18n from 'i18next'
import languages from './strings'

i18n
  .init({
    lng: 'en', // set dynamically on build
    // lngs: Object.keys(languages),
    fallbackLng: 'en',

    ns: ['common'],
    defaultNS: 'common',

    debug: true,

    interpolation: {
      escapeValue: false // not needed for react!!
    }
  })

Object.keys(languages).map(lang => {
  i18n.addResourceBundle(lang, 'common', languages[lang], true)
})

export default i18n
```

strings/index.js
```javascript
import en from './en.json'
import de from './de.json'

export { en }
export { de }

export default { en, de }
```

strings/en.json
```json
{
  "hello_world": "Hello World"
}
```

And I'm using the json-loader with webpack.

**Comment by sorahn** [comment]
FWIW This only seems to happen when saving the file with the decorator in it.

**Comment by jamuhl** [comment]
@sorahn can you try to exclude resources from json-loader?

**Comment by sorahn** [comment]
@jamuhl do you mean make the strings .js files and objects?

```
      { test: /\.json$/, loader: 'json-loader', exclude: /node_modules/ },
```

**Comment by jamuhl** [comment]
i wouldn't bundle the translations (or i would only bundle them in cases where you don't have a backend server -> standalone apps on mobiles). I would use the https://github.com/i18next/i18next-xhr-backend and let i18next load that json files from server.

**Comment by sorahn** [comment]
@jamuhl Yeah, we're not going to have a backend available to us, so we're going to just bake all 8 languages into the bundle.  This bit loops through them and adds them all in.  
```
Object.keys(languages).map(lang => {
  i18n.addResourceBundle(lang, 'common', languages[lang], true)
})
```

Everything itself is working great, the only issue comes from saving a file with the @translate decorator.  When the HMR injects the newly created class, it doesn't have it's decorator function available.

**Comment by jamuhl** [comment]
Tried to reproduce:

added webpack-dev-server to example: https://github.com/i18next/react-i18next/tree/master/example#run-with-webpack-dev-server

added a class component with translate hoc as decorator: https://github.com/i18next/react-i18next/blob/master/example/app/components/View.js#L6

doing changes there all works well...can you provide a sample to reproduce this?

**Comment by jamuhl** [comment]
only difference i can see is i use a backend in the sample...but how that impacts the decorator to not get executed proper...idk

**Comment by sorahn** [comment]
We've got it at work right now, Let me try and strip it down and see if it still reproduces the problem.  And I'll get it pushed somewhere.

**Comment by LucaColonnello** [comment]
Hey guys,
there's a solution.

Put this after i18next.init call.

```js
if (module.hot) {
  module.hot.accept([assetsModule], () => {
    const res = require([assetsModule]);
    Object
      .keys(res)
      .forEach((lang) => {
        Object
          .keys(res[lang])
          .forEach((namespace) => {
            i18next.addResourceBundle(lang, namespace, res[lang][namespace], true, true );
          })
        ;
      })
    ;

    i18next.emit('loaded');
  });
}
```

Where `[assetsModule]` is the module that export the resources, used as resources in the initial option and also for the hot replacement.
You can also use ([`i18next-resource-store-loader`](https://www.npmjs.com/package/i18next-resource-store-loader)).


### Complete Example:

```js
import i18next from 'i18next'
const resources = require("i18next-resource-store-loader!../assets/i18n/index.js");

i18next
  .init({
    lng: 'en', // set dynamically on build
    fallbackLng: 'en',

    resources: resources,
    debug: true,

    interpolation: {
      escapeValue: false // not needed for react!!
    }
  });

if (module.hot) {
  module.hot.accept("i18next-resource-store-loader!../assets/i18n/index.js", () => {
    const res = require("i18next-resource-store-loader!../assets/i18n/index.js");
    Object
      .keys(res)
      .forEach((lang) => {
        Object
          .keys(res[lang])
          .forEach((namespace) => {
            i18next.addResourceBundle(lang, namespace, res[lang][namespace], true, true );
          })
        ;
      })
    ;

    i18next.emit('loaded');
  });
}
```


With this, every time you change a value into resources, the hot reloader replace all the namespaces with new values into i18next and emit `loaded` event.

Doing so, all the react component wrapped with the `Translate` component, will be re-rendered.

**Comment by jamuhl** [comment]
@LucaColonnello cool...if you don't mind i will add i18next-resource-store-loader to i18next.com under ecosystem...so others can find this too.

**Comment by LucaColonnello** [comment]
Yeah @jamuhl , I think it could be helpful to also add the solution I posted for the hot reloading in a FAQ or How To section..

Some users use hot reloading (all users indeed) and an Internationalization system has to manage all the users' needs.

**Comment by jamuhl** [comment]
@LucaColonnello agree. At our place we just have the translations on the server and use the xhr backend coming with i18next. That configuration works with hotreload too. Things only do not work when you start to require the translations and bundle them into build output.

so i will add some writeup on both alternatives. or we might even add another sample to the example folder in this repo (one using xhr backend is already done).

**Comment by LucaColonnello** [comment]
@jamuhl really good!!!!!

**Comment by Ganasist** [comment]
I'm using this great package and it's working fine, love it. I have one weird scenario...

Normally I use the @translate(['namespace']) decorator in ALL of my components, and it's been working fine. However, for one component (that is identical and in parallel with many others), this does not work. In all of the parallel components I import translate and use the decorate as above, I get 't' in the props and use it as usual...

BUT, I'm able to fix this one rogue component by reverting to this format..

export default translate(['login'])(LoginForm)

...and skipping the decorate pattern altogether. Any idea why this might be? Here's the error in the one component when I try to use the standard decorator approach (and which I never see in any other component):

Uncaught TypeError: Cannot read property 'loadNamespaces' of undefined - translate.js:55

**Comment by jamuhl** [comment]
@Ganasist seems like in that case somehow context.i18n is not set https://github.com/i18next/react-i18next/blob/master/src/translate.js#L14 ...but honestly no idea why

**Comment by jamuhl** [comment]
closing this...reopen if still having some issues..

**Comment by hollanderbart** [comment]
I also have the same issue as @Ganasist. Get the error `Uncaught TypeError: Cannot read property 'loadNamespaces' of undefined - translate.js:55`
when I use the HOC translate method.

export default translate('common')(ComponentName)

When I don't include translate and only use the i18n, everything goes fine.

**Comment by jamuhl** [comment]
@hollanderbart still no idea how to track this one down....

**Comment by hollanderbart** [comment]
I'm getting this error when using Jest as my unit test framework. When I use the following setup for my i18next (HOC): 

`import { translate } from 'react-i18next'
...
  render() {
    const { t } = this.props
return(
<div>
<input type="text" value={this.state.filter} placeholder={t('common:filter.filterpalceholder')} onChange={this.handleFilterChange}></input>
</div>
)
}
export default translate('common')(FilterRow)
`

I get this error in my Jest unit test:
`- TypeError: Cannot read property 'loadNamespaces' of undefined
        at Translate.componentWillMount (node_modules/react-i18next/dist/commonjs/translate.js:57:15)`

When I would replace my code to the direct i18n setup, with no import { translate }, everything goes perfect.

**Comment by jamuhl** [comment]
Cannot read property 'loadNamespaces' of undefined 

there is no i18n on context --> that's what i18nextProvider does...for tests i would mock away the translated hoc...

or better not test the extended component  but only the component:

export myComponent function(props) {...} <- test this one
export default translated('ns')(myComponent); <- not that

**Comment by hollanderbart** [comment]
I agree with you. The only this is that if I unmock translate using `jest.unmock('react-i18next');` I keep getting the ` TypeError: Cannot read property 'loadNamespaces' of undefined` error

**Comment by hollanderbart** [comment]
Thanks. It's working now.

**Comment by hollanderbart** [comment]
I got it working. Finally.
Via this post: https://github.com/i18next/react-i18next/issues/3, I was mentioned this post:

**Comment by gnapse** [comment]
> we just have the translations on the server and use the xhr backend coming with i18next. That configuration works with hotreload too.

@jamuhl I have a react app created with create-react-app, we're also using i18next with the xhr backend, and I'm starting to use HMR, which works fine for components, but when I change the translation files in the `public/` directory, I'm not sure how I can make it work. Can you share what `module.hot.accept(...)` call you're making to make this work?

**Comment by gnapse** [comment]
Thanks, I may indeed take this to SO. Just one last question: it has always seem weird to me that even when the translation files in the `public/` directory are not part of the bundle, when I change them, the app running in dev mode automatically reloads the whole page, exactly as when I change something in the code inside `src/`.

**Comment by jamuhl** [comment]
hm...might trigger a reload in that case - than it's just the question of what is coming in `module.hot.accept("i?!?!?"` and from there trigger a reloadResources:

```
if (module.hot) {
  module.hot.accept("i?!?!?", () => {
   // then just
   i18next.reloadResources();
 });
}
```

If you got more feedback fro SO - i would be happy if you share your solution - would be a great add to documentation.

**Comment by gnapse** [comment]
Thanks for the `i18next.reloadResources` tip! That was a missing piece in my puzzle. I've yet to unravel if this is going through `module.hot` at all. If I find anything relevant I'll bring it back here.

**Comment by LucaColonnello** [comment]
Have you already tried the solution I've posted above?

**Comment by gnapse** [comment]
@LucaColonnello I believe your solution is not meant for when the XHR backend is being used. Or is it? For starters, what should I use in place of `[assetsModule]`? You say it is this:

> Where [assetsModule] is the module that export the resources, used as resources in the initial option and also for the hot replacement.

But what module you refer to that export what resources? If you mean modules exporting your translations, when you use the XHR backend, you don't keep the translations in the bundle. They're loaded via ajax request when the app is loaded.
